### PR TITLE
[RAPPS] Minimal UI mode for AppWiz ...

### PR DESCRIPTION
**Pull request by HBelusca**
... aka. make @cbialorucki happy :joy: 

## Purpose

Remove unneeded stuff from RAPPS UI when it is run in "App-Wizard" (i.e. "Add/Remove Programs") mode.
You can test it via `rapps /appwiz` and, in ReactOS, by also opening the "Add/Remove Programs" applet from the control panel (i.e. running `appwiz.cpl`).
This is in a sense, a follow-up to PR #6319 , see comments: https://github.com/reactos/reactos/pull/6319#issuecomment-1880879751 , https://github.com/reactos/reactos/pull/6319#issuecomment-1880972571 .

_**RAPPS in AppWiz mode**_
image

_**RAPPS in normal mode**_
image

## TODO / Questions for reviewers

- [x] _**~~(NOTE: I have no idea why the toolbar doesn't correctly show the button labels unless the window is enlarged quite a bit...)~~**_ Now fixed with the last commit.
- [x] Remove some unneeded stuff from the PR.
- [x] ~~Do you think the `[SDK:UI] Add a CToolbar::DeleteButtonByCmd() method` could be of any use in the future?~~ Removed from this PR.
- [x] There are two `#if 0`'ed code blocks in the PR, as alternative places where the code could be implemented. ~~What do you think about these, compared to where the actually-compiled code is?~~ Addressed by moving this code into the `WM_CREATE`, where it makes more sense, since the deletion of menu items and toolbar buttons has to be done just once.
- [x] ~~What do you think about the commit `[RAPPS] Allow multiple RAPPS windows when it's started in AppWiz mode.` ?~~ Will be addressed differently in a separate PR.
- [x] About the `[RAPPS] Remove the IDR_APPLICATIONMENU menu, redundant [...]` commit: ~~What do you think about the alternative, namely: keeping this menu separate while removing the identical one that is in the main menu? In that case, how can we make the main menu display that submenu correctly when clicking on the "Programs" menu item?~~ The current way of dealing with it is simpler than the hypothetical alternative.

### CODE REVIEW

**Comment by JoachimHenze** [comment]
I do like the idea of this PR very much and therefore I do approve it already.

Out of interest: Can you compare the binary size of a GCC dbg build of rapps.exe before and after your PR?
(I am interested to see whether the stripped rc parts do outweight the added source-code or not).

Ftr: I would have liked it even more if a minimal offline-version of rapps would have just been compiled directly into appwiz.cpl. E.g. by digging out some ancient code from SVN where Rapps was still pure C.
E.g. the state of roughly SVN70255_2015-12-03 would allow to create a complete offline version of that app with just about ~ 100kilobyte of binary size.
[rapps_SVN70255_2015-12-03.zip](https://github.com/reactos/reactos/files/14726744/rapps_SVN70255_2015-12-03.zip)
And that could be used then even if the user decided to delete rapps.exe entirely, and just stick with the Windows-stock-like appwiz.cpl.
But that's just added as a thought here.

**Review by JoachimHenze — APPROVED** [review]
I do like the idea of this PR very much and therefore I do approve it already.

Ftr: I would have liked it even more if a minimal offline-version of rapps would have just been compiled directly into appwiz.cpl. E.g. by digging out some ancient code from SVN where Rapps was still pure C.
E.g. the state of roughly SVN70255_2015-12-03 would allow to create a complete offline version of that app with just about ~ 100kilobyte of binary size.
[rapps_SVN70255_2015-12-03.zip](https://github.com/reactos/reactos/files/14726744/rapps_SVN70255_2015-12-03.zip)
And that could be used then even if the user decided to delete rapps.exe entirely, and just stick with the Windows-stock-like appwiz.cpl.
But that's just added as a thought here. Your PR doesn't make my idea impossible or harder to do in any way.

File: base/applications/rapps/lang/it-IT.rc [hunk 1]
```diff
@@ -45,22 +45,6 @@ BEGIN
...
-        MENUITEM "&Modifica", ID_MODIFY
-        MENUITEM SEPARATOR
-        MENUITEM "&Cancella dal registro", ID_REGREMOVE
-        MENUITEM SEPARATOR
-        MENUITEM "&Aggiorna\tF5", ID_REFRESH
-        MENUITEM SEPARATOR
-        MENUITEM "Aggiorna Da&tabase\tCtrl+F5", ID_RESETDB
-    END
-END
-
```
Comment (JoachimHenze) [inline base/applications/rapps/lang/it-IT.rc:63]: Just a question: That cleanup part of your PR of removing IDR_APPLICATIONMENU in all rcs... Is that possible also without the rest of your PR?

Comment (HBelusca) [reply base/applications/rapps/lang/it-IT.rc:63]: Yes, this removal is in a separate commit that can compile without the rest.

**Comment by cbialorucki** [comment]
This PR helps the uninstall experience immensely. However, I still have a few issues with it.

1. I have to expand the window to be very wide to get the text for each button to display. The text for each button should always be displayed unless the window is very thin. Here's a picture of the minimum width I need for the window before it starts displaying the button text in English; any thinner and the text next to the buttons are not displayed.
image
Perhaps this would be easier to do if we added a "View" menu to the menu bar with the 'Details' and 'List' view options instead of using the combobox to the left of the search box.

2. I would like to see the tree view to the left be completely removed. It's distracting and unnecessary.

3. Since RAPPS is launched anyways, an option to return to the full RAPPS interface would be nice. This could be added under the proposed view menu or in the file menu.

Overall though this is a huge step in the right direction, and I can add these features in a future PR if need be. Great work!

**Comment by HBelusca** [comment]
@cbialorucki Some thoughts about your different points:
1. This is indeed what I observe too (see my note). There's a call somewhere to readjust the toolbar size, maybe it's what's missing. I'll try to investigate further.
2. Maybe, but currently it helps filtering things (and is useful on windows for example): "Installed" shows everything, "Applications" just shows the applications and "Updates" just shows the OS/etc. updates (which is, granted, empty in ReactOS, but not on Windows for example). And I was thinking, that if we also get support for installing separate ReactOS components, we could also have a section in the tree for these.
3. Interesting proposal to be thought about, but for another PR.

**Comment by whindsaks** [comment]
I've been playing with the idea of properly implementing the uninstall stuff inside appwiz.cpl.

- Just a simple listbox like Windows 95
- NT5 style but perhaps just a ListView
- IShellFolder based like Vista

**Comment by cbialorucki** [comment]
@HBelusca 

1. Thanks for looking into it, displaying the text for the buttons is very important for understanding what each button does, especially with our current icon set.

2. No version of Windows offers application updates through appwiz.cpl., nor would I think to open appwiz.cpl to update a program. I have never wanted to filter out the programs in appwiz.cpl and would find this quite annoying. I do like your proposal about the "Add/Remove components" feature but the tree is the wrong control for this option. This option could be put in the File menu or as a link to the side; not in the tree.

3. Understandably so, just something to think about.

**Comment by HBelusca** [comment]
> I've been playing with the idea of properly implementing the uninstall stuff inside appwiz.cpl.
> 
>     * Just a simple listbox like Windows 95
> 
>     * NT5 style but perhaps just a ListView
> 
>     * IShellFolder based like Vista

Could be an interesting alternative (and offering a button to launch RAPPS instead).
For reference, how it looks in Wine:
image

**Comment by HBelusca** [comment]
@cbialorucki About point 2 (the updates):
It's not about "offering" updates, it's just about showing _which ones are installed on the system_ (be it OS updates or application updates).
Windows has such a thing too.

Win2k/XP/2k3:
image_winXP

Win7:
image_win7

**Comment by cbialorucki** [comment]
@HBelusca 

I get what you're saying now, thanks for responding cordially. I do still think the tree on the righthand side is the wrong control for this but that is outside the scope of this PR.

For the record, I also really don't like how Wine does it; what you put together looks better.

**Comment by HBelusca** [comment]
Toolbar text buttons showing/hiding during resize is fixed with last commit.
Note: a nice feature could be to have the buttons text progressively showing/hiding during resize (when reducing window size: hiding text from the rightmost button and going back to the first leftmost button ; when increasing window size: showing text from the leftmost button and going forth to the rightmost button).

**Comment by JoachimHenze** [comment]
I do also like very much the visuals that Hermes added here for the Minimal UI. Whether the treeview should be part of the pure appwiz.cpl is a matter of taste (I do actually like its visuals) and size (I do appreciate everything that is small). We could start without it, and see how far it will get us. OS updates on ros won't be a thing for the next 20 years most likely. updates-question is only interesting when managing them on windows. i could live also with accessing that just via the file menu (as that would result in smaller binary most likely than using the treeview).


`Do you think the [SDK:UI] Add a CToolbar::DeleteButtonByCmd() method could be of any use in the future?`
Yes, looks helpful.


`What do you think about the commit [RAPPS] Allow multiple RAPPS windows when it's started in AppWiz mode. ?`

+4 lines of code isn't much to offer something like that. Interesting. But even if it is just 4 lines, I do see absolutely no real world usecase for that.


`What do you think about the alternative, namely: keeping this menu separate while removing the identical one that is in the main menu? In that case, how can we make the main menu display that submenu correctly when clicking on the "Programs" menu item?
`

Unsure. For now I do only think every avoided duplication is a good thing. Not thinking much more than that. What would be the benefit of removing the other one **in your eyes**, that you do consider it?

**Comment by HBelusca** [comment]
>  +4 lines of code isn't much to offer something like that. Interesting. But even if it is just 4 lines, I do see absolutely no real world usecase for that.

These "+4" lines of code are part of the "decoration" (a line for a comment, two lines for opening/closing brackets for an if block, and one actual line of code that is the if condition itself).
When I'm asking whether the commit is needed/useful, I'm talking about the suggested functionality. Not the fact it adds or remove lines of code.

**Comment by learn-more** [comment]
Considering this PR is created after #6652, I'd say it is only fair to merge #6652 first.

**Comment by HBelusca** [comment]
> Considering this PR is created after #6652, I'd say it is only fair to merge #6652 first.

Does #6652 intersect with the modifications brought here?

**Comment by whindsaks** [comment]
> Does #6652 intersect with the modifications brought here?

I doubt it, it does not touch anything UI related.

**Comment by JoachimHenze** [comment]
@HBelusca please resolve the conflicts, so we can merge that nice PR of yours.

**Comment by HBelusca** [comment]
Conflicts resolved, let's proceed with some more feedback from reviewers, and in the meantime I'll try to implement more UI minimalization based on @cbialorucki last comments.

**Review by oleg-dubinskiy — APPROVED** [review]
Actually a nice idea. I'm up for it too. And Joachim's suggestion looks OK for me too, and I'm agreed with moving the most often used commands from right-click menu to the toolbar buttons (like Remove From Registry) (and on the contrary, move less needed ones to the right click menu). But I'd keep Refresh toolbar button as-is, because refreshing the list via F5 is not enough obvious personally for me (same as for some other people I'm sure).
Regarding your questions, I have the following opinions:
> Do you think the [SDK:UI] Add a CToolbar::DeleteButtonByCmd() method could be of any use in the future?

Can you please add a short explanation of what this method actually does? It removes the button from the toolbar? Or from somewhere else? Or... ? :thinking: It's not obvious enough from its name for me at least. But if this is what I'm thinking about, then yes, it could be useful, if there is no any other way to avoid initialization of some buttons when it's not needed, e. g. by doing that only for appwiz mode and don't otherwise (use an appropriate if-else condition if possible). Otherwise, if it's possible, that method simply makes no sense.
> There are two #if 0'ed code blocks in the PR, as alternative places where the code could be implemented. What do you think about these, compared to where the actually-compiled code is?

If they are just another places for the same code already added and they don't add any new functionality, then it makes no difference where to add it, if it will be called from the same place anyway. Move it there if you wish or if it actually looks better in those places.
> What do you think about the commit [RAPPS] Allow multiple RAPPS windows when it's started in AppWiz mode. ?

I'd totally agree with Joachim here. Is it useful in any way at all? How many real usage cases can it have? I think none. It makes sense for the apps those do some task that might be done multiple times at once. But in our case, if user wants to uninstall one app at once (might he need to uninstall two or more apps in the same time?), he doesn't need to run multiple instances of Rapps.
> About the [RAPPS] Remove the IDR_APPLICATIONMENU menu, redundant [...] commit: What do you think about the alternative, namely: keeping this menu separate while removing the identical one that is in the main menu? In that case, how can we make the main menu display that submenu correctly when clicking on the "Programs" menu item?

I also don't think it makes sense at all. It's more like stylish-only thing depending on your (or anybody else) personal taste. But from the technical point of view, it changes nothing.

**Review by JoachimHenze — APPROVED** [review]
Approved. I wouldn't mind if we merged this PR immediately, and would do further improvements/simplifications in this context in a follow-up-PR. Thx for considering the suggestions.

**Comment by HBelusca** [comment]
Here is the current "reasonable" proposal (for this PR) for the APPWIZ-mode of RAPPS, mainly adapted from @JoachimHenze 's screenshot and other discussion:
last_proposal_for_this_PR

1. The TreeView stays, as it allows to just continue using existing code. (Removing the treeview and using buttons or stuff would instead require more new code to be added.)
2. The menu bar stays for now. The "File" menu only contains "Quit" for the time being (the "Settings" item is removed since it pertains only to RAPPS in general, and not to the APPWIZ-mode. In the future we could add there what @cbialorucki proposed in point 3 of [his feedback](https://github.com/reactos/reactos/pull/6655#issuecomment-2015759341).
Note that for the time being, if you want to get RAPPS back in normal mode, you can just open it as usual: it can be used in parallel to the one running in APPWIZ-mode.
3. I keep the toolbar buttons as I suggested in the past; in particular I don't add a "Remove from Registry" button here because I consider it as an "advanced-user" command, that I don't want to have readily visible there.
4. I keep the "Refresh" button there, see @oleg-dubinskiy's comment.
5. The "Selected: ..." count in the status bar is removed in APPWIZ-mode, because it doesn't make sense here. (This count is for the number of selected apps waiting to be downloaded and installed.)
6. The "Description" column is, for now, removed _**for installed apps only**_ (in both APPWIZ and regular mode), because for them it currently doesn't bring that much of information. In the future, we could add more columns there ("Publisher", "Size", "Installed on", ...) and the possibility to show/hide columns, and then, re-display optionally the "Description" column as well.
7. For @JoachimHenze : the trailing "column" (which is not a column) in the ListView is standard and will appear as long as the sum of widths of all the columns present is (strictly) less than the width of the ListView. You can observe that with the taskmgr, or the shell (in Report view) in ReactOS and Windows as well.
8. The "Welcome" text shown in the lower panel, when opening RAPPS or a treeview category is selected (without selecting an application in the listview), is changed to display a suitable explanatory text in APPWIZ-mode.

**Comment by HBelusca** [comment]
> What do you think about the commit `[RAPPS] Allow multiple RAPPS windows when it's started in AppWiz mode.` ?

The original purpose of that commit was mainly in order to allow an RAPPS in APPWIZ-mode to run alongside an RAPPS running in standard mode.
The problem is that as the commit is currently made, it also allows _multiple_ RAPPS instances all running in APPWIZ-mode. (And both @JoachimHenze and @oleg-dubinskiy are indeed dubious about the usefulness of that latter point in particular.)
I've checked on both Windows 2000 and 2003, and there, the appwiz window is also single-instanced. So a bit more work is needed on my part in order to allow what I originally wanted to do ("original purpose").

**Comment by HBelusca** [comment]
> There are two `#if 0`'ed code blocks in the PR, as alternative places where the code could be implemented. What do you think about these, compared to where the actually-compiled code is?

I need code-reviewers to provide me with feedback about where the best place of doing this is.
* The two `#if 0`-ed blocks are in [appview.cpp-L1547](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R1547) and [appview.cpp-L2002](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R2002);
* The current enabled code is in [appview.cpp-L2043](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R2043)

**Review by Doug-Lyons — APPROVED** [review]

**Comment by JoachimHenze** [comment]
> I need code-reviewers to provide me with feedback about where the best place of doing this can be. 
> * The two `#if 0`-ed blocks are in [appview.cpp-L1547](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R1547) and [appview.cpp-L2002](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R2002);
> * The current enabled code is in [appview.cpp-L2043](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R2043)

If it would be possible to pack all that `if (m_MainWindow->m_bAppwizMode)` conditional code at just a single location, that would be optimal of course. Personally I do think the [appview.cpp-L2043](https://github.com/reactos/reactos/pull/6655/files#diff-7efb93629c9d0350877e09999dad2d28f1042465b2bd0a198ed8cc930f320ea4R2043) position in function CApplicationView::SetDisplayAppType() would sound most natural to me of all 3 mentioned places.

**If fusing all 3 to just there would not be possible** for whatever reason, then I do think limiting their count to 2 locations instead of 3 would be desireable. If their place is arbitrary (when having 2 blocks just) then funcname CApplicationView::Create(HWND hwndParent) sounds like a more logical choice to me than CApplicationView::ProcessWindowMessage().

**But I don't have a strong preference for any of those**.
To me reducing the number of places to touch is more important than the "where-stuff-is-done".
Go with what works best for you and feels flexible.

**Comment by HBelusca** [comment]
@JoachimHenze Regarding this last point, I moved this code into the `WM_CREATE` handler of the appview.

File: base/applications/rapps/appview.cpp [hunk 2]
```diff
@@ -1359,16 +1411,18 @@ CAppsListView::SetDisplayAppType(APPLICATION_VIEW_TYPE AppType)
 
             /* Add columns to ListView */
             szText.LoadStringW(IDS_APP_NAME);
-            AddColumn(ColumnCount++, szText, 250, LVCFMT_LEFT);
+            AddColumn(ColumnCount++, szText, 368 /*250*/, LVCFMT_LEFT);
```
Comment (JoachimHenze) [inline base/applications/rapps/appview.cpp:1414]: That's what a version control system is for. Let's keep the heads sources clean. Rest of PR LGTM.
```suggestion
            AddColumn(ColumnCount++, szText, 368, LVCFMT_LEFT);
```

**Pull request merged**